packages/util/src/error.ts

```diff
@@ -6,3 +13,43 @@ export function errorWithCause(message: string, cause: Error): Error {
   }
   return newError;
 }
+
+/**
+ * Represents an error that was caused by another error.
+ */
+export class ErrorWithCause extends Error {
+  /**
+   * Create an instance of ErrorWithCause.
+   * @param message Error message.
+   * @param cause Original error, causing this error.
+   */
+  constructor(message: string, public cause: Error) {
+    // There is an issue with the prototype chain when extending from Error: https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-2.html#support-for-newtarget
+    super(message); // 'Error' breaks prototype chain here
+    Object.setPrototypeOf(this, new.target.prototype); // restore prototype chain
+    this.name = 'ErrorWithCause';
+
+    // Stack is a non-standard property according to https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error#Custom_Error_Types
+    if (this.stack && cause?.stack) {
+      this.stack = `${this.stack}\nCaused by:\n${cause.stack}`;
+    }
+  }
+
+  /**
+   * Root cause of the error.
+   * If there are multiple errors caused one by another, the root cause is the first error that occurred.
+   * In case there is no root cause.
+   */
+  get rootCause(): Error {
+    return isErrorWithCause(this.cause) ? this.cause.rootCause : this.cause;
```

Comment:
Minor: will run in a loop if we create a cyclic chain. But unlikely.

Reply:
Question: Does JSON.stringify also evaluate the getter? This would be a nice in case somebody serializes such an error e.g. in a logger this will happen at a certain point right? Or how does our provided winston logger hanndle a `logger.error(err)` for this ErrorWithCause and some chain of errors inside.

Reply:
> Minor: will run in a loop if we create a cyclic chain. But unlikely.

Good point, this is the same behaviour as in the old errorWithCause function and it should probably never happen.
The only way I can think of achieving this would be:
```ts
const cause = new ErrorWithCause('cause', new Error('root cause'));
const error = new ErrorWithCause('error', cause);
// Creates circular chain
error.cause = cause;
```

Reply:
> Question: Does JSON.stringify also evaluate the getter? This would be a nice in case somebody serializes such an error e.g. in a logger this will happen at a certain point right? Or how does our provided winston logger hanndle a `logger.error(err)` for this ErrorWithCause and some chain of errors inside.

I checked and those are two separate issues I think.

1. JSON.stringify gives me:
```ts
JSON.stringify(
    new ErrorWithCause('ErrorWithCause', new Error('root cause')),
    null,
    2
  )
// result
// {
//   "cause": {},
//   "name": "ErrorWithCause"
// }
```

There are two aspects to it.
* a) Yes, there is no rootCause on it, I could add it.
* b) Can you see what JSON.stringify does for Error? -.-
  An example:
  ```ts
  JSON.stringify(new Error('root cause'), null, 2)
  // result
  // {}
  ```
  Not sure whether we should do anything about this. On the one hand it's nice (and not difficult to implement) - on the other hand the behaviour is different from the standard error. Would you call JSON.stringify on all errors (Error and ErrorWithCause) you would get different results.

2. What does the logger do? It simply prints the error message. So for `new Error('root cause')` that would be `'root cause'`. This is the same `errorWithCause` and `ErrorWithCause`, e. g. `new ErrorWithCause('the error', new Error('root cause'))` results in `'the error'`. I guess that is not optimal, it might make sense to improve that, but I don't think this is related to this PR.

Reply:
Ok fine for me. Perhaps we make a item to the backlog to improve the logging if our `ErrorWithCause` is logged by our loggers.

Reply:
For the cyclic thing I would wait until this actually happens.

Reply:
Or even if error are logged for that matter...